### DISCUSSION TITLE
Viewing HTTP content

### DISCUSSION THREAD

**Discussion by aplatypus** [testingisdocumenting/webtau#1132] [Q&A]
**HTTP information**

I'm not sure how 'discussions' work.  I want a new topic under the "WebTau Output" discussion, only that made it look like I was answering another comment.    So this is a new topic.

I have sent a request to a Swagger service and it returned the error: "`selenium.WebDriverException: Invalid Host header localhost:31927`".

* I think that I know the problem this time.  However, I went looking for some kind of view of the HTTP Request packet sent and the returned Response packet.  
* The Swagger, `REST_ENABLE_SWAGGER_INTERFACE` web page (for example)  shows what has been sent and returned on the browser Swager API page when you do a test call.  
* I suggest that something for testing web calls and the like ought to have ways to "report" on HTTP request and response  data, for quick turnaround on these simple error types.
* Perhaps this facility is already present(???)

**Comment by MykolaGolubyev** [comment]
If you plan to test HTTP like REST you need to use http module and not browser. 
Could you please provide more info on what are you trying to achieve?

Not sure how to properly use Discussion. What you do works for me

**Comment by aplatypus** [comment]
Hi Mykola ... brief outline for the inital tests.  I want to get some simple tests working so I can automate thing we need to do manually with things like Postman.  Yes, want to use REST over HTTP.

- Target is `localhost`
- It needs an authentication session cookie.
   - The esiest thing to do is show you the CURL command I use from the shell.
   - `curl -X GET "http://localhost:10007/api/v1/caa/list/South" -H  "accept: application/json" -H  "Cookie: session=560793879503xxxxx31"`
   - Initially I was asking, "How to set a cookie?"  I am using the `http` module to issue the request. 
- What I'm not clear on is how to pass the cookie as shown on the command line above?
-  At present this example falls over in Groovy with an exception:  
```
        HttpHeader sessionHeader = WebTauDsl.http.header(
            "Accept" , "application/json",
            "Cookie" , "session=5607938795038562531"
        );
        
        WebTauDsl.http.get( "http://localhost:10007/caa/api/v1/caa/list/"+ boardName, sessionHeader ) {
            String result = body.get("caaList");
            logger.info("${FUNC} result => ${result}.");
        };
```

 - exception:  
```
groovy.lang.GroovyRuntimeException: Ambiguous method overloading for method org.testingisdocumenting.webtau.http.Http#get.
Cannot resolve which method to invoke for [class java.lang.String, class org.testingisdocumenting.webtau.http.HttpHeader, class au.com.fourtel.fourams.controller.tmacs.CaaMonitorTest02$_testListCaas_closure1] due to overlapping prototypes between:
	[class java.lang.String, class org.testingisdocumenting.webtau.http.HttpHeader, interface org.testingisdocumenting.webtau.http.validation.HttpResponseValidatorWithReturn]
	[class java.lang.String, class org.testingisdocumenting.webtau.http.HttpHeader, interface org.testingisdocumenting.webtau.http.validation.HttpResponseValidator]
	at groovy.lang.MetaClassImpl.doChooseMostSpecificParams(MetaClassImpl.java:3426)
	at groovy.lang.MetaClassImpl.chooseMostSpecificParams(MetaClassImpl.java:3402)
```
  - I can see what is it confused over, but I don't think it "should" be ambiguous
- First though is that a suitable way to set-up my REST call now?

**Reply by MykolaGolubyev** [reply]
Not at my computer at the moment.
From the first look it seems you are using java syntax in groovy file with some mixed approach.

Take a look at CRUD example documentation.

Also I see you use get on the body but assign it to String. Check the DataNode page docs to understand types WebTau operates with.

**Reply by aplatypus** [reply]
* Eventually I need to use Java.  Fortunately since I do not need the `Browser` I can convert this test  set back to a Java file.
  * Running Java over the call seems to be fine.
  * I took a close look at the two interfaces the Groovy runtime complained about.  They only differ in return type. 
* With a Groovy approach, Groovy give me that error; no matter how I changed the syntax around.  The example above is after many variations based on the REST examples. First for Groovy and later I tried a Java like variation because Java is better at strict typing.
  * I'm happier with the Java test file.  There may still be an ambiguity as far as the Groovy runtime goes.  
  * In Java method signatures must differ by parameter types.  A method can't be overloaded just by changing the return type.  Groovy clearly does a good job, but I can see how it found those two methods ambiguous.  
  * One way out of it is to fully qualify the method you want using the package path not just the name.
  * May be I'll get time to give a Groovy version another go, now that I have a working example.

**Reply by MykolaGolubyev** [reply]
@aplatypus I am not able to reproduce using `Groovy`. What version of `WebTau` do you use?

**Reply by MykolaGolubyev** [reply]
Here are additional tests in a PR I just added: https://github.com/testingisdocumenting/webtau/pull/1150/files
